### PR TITLE
Added dismiss upon selection to product 'Order By'

### LINKED ISSUE #4496: Products Bottom Sheet Should be Dismissed After Choosing an Option
### Expectations

The Products List' Sort and Filter bottom sheets should be dismissed right away once you pick an option. 

### Actual

The Products List' Sort and Filter bottom sheets are not dismissed once you pick an option. 

https://user-images.githubusercontent.com/198826/123838807-a26cd680-d8c9-11eb-9455-cb6e7effc6c7.MP4

### To Reproduce

1. Navigate to Products List
2. Tap on Sort or Filter
3. Pick a different option. 

The bottom sheet should be dismissed.

### PR DESCRIPTION
**Pull request by iamgabrielma**
Fixes #4496 

## Description

This PR fixes the Products List "Sort by" modal that appears when the button is tapped. Now the modal will disappear after the user has selected one of the available choices.

4496-fix

## Changes

I added a `dismiss()` call at the end of the current `tableView(_:didSelectRowAt:)` delegate method for `BottomSheetListSelectorViewController`, so is triggered after the user selects one of the options.

The original issue mentions the Filter button as well, but this one does not display a modal when tapped, possibly has been changed since the original issue was reported, so this change only affects the "Order by" button.

## Testing steps

1 - Navigate to Products List
2 - Tap on `Sort by` 
3 - Pick a different option.

Update release notes:
- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WooCommerce/Classes/ViewRelated/BottomSheet/ListSelector/BottomSheetListSelectorViewController.swift
Comment (shiki) [inline WooCommerce/Classes/ViewRelated/BottomSheet/ListSelector/BottomSheetListSelectorViewController.swift]: I looked at other places where this ViewController is used. It looks like the original expectation is to have the _presenting view controller_ dismiss the `BottomSheetListSelectorViewController`. Examples:

For the "More details" in a product:

https://github.com/woocommerce/woocommerce-ios/blob/c57b3472b0569f732a8e70728fcf42955cc746c5/WooCommerce/Classes/ViewRelated/Products/Edit%20Product/ProductFormViewController.swift#L565-L591

For the product type selection when adding a product: 

https://github.com/woocommerce/woocommerce-ios/blob/c57b3472b0569f732a8e70728fcf42955cc746c5/WooCommerce/Classes/ViewRelated/Products/Add%20Product/AddProductCoordinator.swift#L46-L53

Notice the `dismiss()` calls in those places. And there are a couple more that I have not listed.

Adding this dismissal here would negate those implementations. Or I don't know what else would happen if `dismiss()` is called twice. While we could remove those places where `dismiss()` is called, I think it's safer to follow the convention instead and manually call `dismiss()` where the sorting dialog is presented. 

To do that, it looks like we can change the implementation of [`ProductsSortOrderBottomSheetListSelectorCommand`](https://github.com/woocommerce/woocommerce-ios/blob/c57b3472b0569f732a8e70728fcf42955cc746c5/WooCommerce/Classes/ViewRelated/Products/BottomSheetListSelector/ProductsSortOrderBottomSheetListSelectorCommand.swift#L22) so it's more like [`ProductTypeBottomSheetListSelectorCommand`](https://github.com/woocommerce/woocommerce-ios/blob/c57b3472b0569f732a8e70728fcf42955cc746c5/WooCommerce/Classes/ViewRelated/Products/Edit%20Product/BottomSheetListSelector/ProductTypeBottomSheetListSelectorCommand.swift#L132). 

Notice that `ProductTypeBottomSheetListSelectorCommand` propagates the `onSelection` event during the `handleSelectedChange` event. That `onSelection` block dismisses the view. We can follow the same strategy with sorting. 

What do you think?

**Review by shiki — CHANGES_REQUESTED** [review]
Hi, @iamgabrielma! Thanks for fixing this. 

This works but I'm worried that the implementation bypasses the current conventions of using the `*Command` classes to dismiss the dialog. I fear that this would cause some hidden regressions in the long run and it might be better to copy the other `*Command` implementations in the code. 

More info inline. 

Let me know what you think.

File: WooCommerce/Classes/ViewRelated/BottomSheet/ListSelector/BottomSheetListSelectorViewController.swift
Comment (iamgabrielma) [reply WooCommerce/Classes/ViewRelated/BottomSheet/ListSelector/BottomSheetListSelectorViewController.swift]: That make a lot of sense, thanks for the pointers and the detailed walkthrough!

**Comment by iamgabrielma** [comment]
The latest commit break some existing unit tests that don't account for the new implementation, I'll review them shortly 👍 

```
/Users/distiller/project/WooCommerce/WooCommerceTests/ViewRelated/Products/BottomSheetListSelector/ProductsSortOrderBottomSheetListSelectorCommandTests.swift:8:89: missing argument for parameter 'onSelection' in call

let command = ProductsSortOrderBottomSheetListSelectorCommand(selected: selected)
        /Users/distiller/project/WooCommerce/WooCommerceTests/ViewRelated/Products/BottomSheetListSelector/ProductsSortOrderBottomSheetListSelectorCommandTests.swift:15:89: missing argument for parameter 'onSelection' in call

let command = ProductsSortOrderBottomSheetListSelectorCommand(selected: selected)
    /Users/distiller/project/WooCommerce/WooCommerceTests/ViewRelated/Products/BottomSheetListSelector/ProductsSortOrderBottomSheetListSelectorCommandTests.swift:29:89: missing argument for parameter 'onSelection' in call

let command = ProductsSortOrderBottomSheetListSelectorCommand(selected: selected)
```

**Comment by shiki** [comment]
No worries! Please press the re-review button if you want me to take a look at this again. 🙂 I looked at the changes and it seems like what we expected! Those darn unit tests!

**Comment by iamgabrielma** [comment]
> No worries! Please press the re-review button if you want me to take a look at this again. 🙂 I looked at the changes and it seems like what we expected! Those darn unit tests!

Great! In the latest changes I changed the implementation of `ProductsSortOrderBottomSheetListSelectorCommand` so it follows the same convention that other bottom sheets behaviour, and dismissal is handled by `onSelection` and `handledSelectedChange` rather than hardcoding the dismissal after the closure.

Regarding the Unit Tests I may need some pointer on how to handle these, as I'm not entirely sure how to test this part of the new implementation for the asynchronous code:

I fixed `testInitialSelectedValue()` as the async code does not affect its initial selected value, but I'm having some troubles with `testSelectedValueAfterChange()` and `testIsSelectedValueAfterChange()`, and I'm not sure if I should change these two to test if the callback is handled correctly instead, as for example [ProductTypeBottomSheetListSelectorCommandTests.swift](https://github.com/woocommerce/woocommerce-ios/blob/fa2bd01933ceb2c1a4e36d434048d134125fad2c/WooCommerce/WooCommerceTests/ViewRelated/Products/BottomSheetListSelector/ProductTypeBottomSheetListSelectorCommandTests.swift) does.

File: WooCommerce/WooCommerceTests/ViewRelated/Products/BottomSheetListSelector/ProductsSortOrderBottomSheetListSelectorCommandTests.swift
Comment (shiki) [inline WooCommerce/WooCommerceTests/ViewRelated/Products/BottomSheetListSelector/ProductsSortOrderBottomSheetListSelectorCommandTests.swift]: We can rename this to clarify it further:

```suggestion
    func test_initial_selected_value_is_stored_in_command() {
```

This is using a very old test method convention. Our new conventions use snake case: https://github.com/woocommerce/woocommerce-ios/blob/develop/docs/naming-conventions.md#test-methods. 

Also, the `testInitialSelectedValue` does not really explain what is _expected_. My suggestion isn't that great but it's probably at least a little bit better than what we have here. 🤪 

I don't think this test is very helpful but it doesn't harm anyone to keep it.

File: WooCommerce/WooCommerceTests/ViewRelated/Products/BottomSheetListSelector/ProductsSortOrderBottomSheetListSelectorCommandTests.swift
Comment (shiki) [inline WooCommerce/WooCommerceTests/ViewRelated/Products/BottomSheetListSelector/ProductsSortOrderBottomSheetListSelectorCommandTests.swift:15]: Since we're only testing that the initial value is set and **we never modify it after that**, a lot of these can be removed. 

```suggestion
        // Arrange
        let selected = ProductsSortOrder.nameAscending

        // Action
        let command = ProductsSortOrderBottomSheetListSelectorCommand(selected: selected) { (selected) in
            // noop
        }

        // Assert
        XCTAssertEqual(command.selected, selected)
```

- The `selectedActions` is not tested so we don't need to keep it
- The `command.handleSelectedChange(selected: .nameAscending)` has [no effect aside from calling the callback](https://github.com/woocommerce/woocommerce-ios/blob/871b9910e5178d39800a11c23af4168f6c55ad8e/WooCommerce/Classes/ViewRelated/Products/BottomSheetListSelector/ProductsSortOrderBottomSheetListSelectorCommand.swift#L46-L48) so it doesn't really affect the value of `command.selected`. 

The shorter the test, the easier it is to understand. 😊

File: WooCommerce/WooCommerceTests/ViewRelated/Products/BottomSheetListSelector/ProductsSortOrderBottomSheetListSelectorCommandTests.swift
Comment (shiki) [inline WooCommerce/WooCommerceTests/ViewRelated/Products/BottomSheetListSelector/ProductsSortOrderBottomSheetListSelectorCommandTests.swift]: It looks like `testIsSelectedValueAfterChange` was originally testing the that `command.isSelected` was returns the correct boolean value for the sort value passed to `command.handleSelectedChange()`. However, `handleSelectedChange` doesn't do anything that affects that since it now just calls a callback function.

We can repurpose this to only test the initial value. Something like this: 

```swift
    func test_isSelected_returns_true_if_given_the_initial_value() {
        // Arrange
        let command = ProductsSortOrderBottomSheetListSelectorCommand(selected: .nameAscending) { (selected) in
            // noop
        }

        // Action
        let isSelected = command.isSelected(model: .nameAscending)

        // Assert
        XCTAssertTrue(isSelected)
    }
```

We can also add another unit test that returns `false` if the incorrect sort value is passed but I leave that up to you if you want to do that. It's not that critical.

File: WooCommerce/WooCommerceTests/ViewRelated/Products/BottomSheetListSelector/ProductsSortOrderBottomSheetListSelectorCommandTests.swift
Comment (shiki) [inline WooCommerce/WooCommerceTests/ViewRelated/Products/BottomSheetListSelector/ProductsSortOrderBottomSheetListSelectorCommandTests.swift]: It looks like the original code is testing that `command.selected` changes whenever `command.handleSelectedChange` is called with a different value. Since `handleSelectedChange`'s behavior changed, this is no longer valid. 

You're right that we can replace this with a unit test that verifies that the callback was called. We can replace this test function with something like this: 

```swift
    func test_handleSelectedChange_notifies_the_callback_the_new_selected_value() {
        // Arrange
        var receivedSelectedSortOrder: ProductsSortOrder?
        let command = ProductsSortOrderBottomSheetListSelectorCommand(selected: .nameAscending) { (selected) in
            receivedSelectedSortOrder = selected
        }

        // Action
        let newSelected = ProductsSortOrder.dateDescending
        command.handleSelectedChange(selected: newSelected)

        // Assert
        XCTAssertEqual(receivedSelectedSortOrder, newSelected)
    }
```

We can have a local variable, `receivedSelectedSortOrder`, that is updated whenever the callback is called. And test that it's updated to what we expect.

**Review by shiki — APPROVED** [review]
Hi, @iamgabrielma! You're right about those tests. They need to be repurposed. I have left suggestions for how we can handle this. Feel free to apply, modify, add until you're satisfied. If you have any questions, let me know. 🙂 

Once the tests pass, feel free to merge. :shipit: But if you'd like me to take a look before merging, that's cool too.

**Pull request merged**
